ui: custom chart selection

vladlos opened the pull request:
fix state of the page when we change the date range

Resolves: #43800

Release note (ui): none

vladlos committed:
ui: custom chart selection

fix state of the page when we change the date range

Resolves: #43800

Release note (ui): none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43913)
<!-- Reviewable:end -->

dhartunian reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43913)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @elkmaster)

___
*[pkg/ui/src/util/query.ts, line 50 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43913#-LyVbuOjBc6gxNiFrb2W:-LyVbuOjBc6gxNiFrb2X:b-7eadjd) ([raw file](https://github.com/cockroachdb/cockroach/blob/568aa83e9bfbd9bbb13006face814d1caa3ff5b0/pkg/ui/src/util/query.ts#L50)):*
> ```XML
> 
> export function removeURLParameters(location: Location, removeParams: string[]) {
>   const deleteRegex = new RegExp(removeParams.join("=|") + "=");
> ```

I'd prefer not to use custom RegExp to manipulate the URL Parameters. The functions above use the `URLSearchParams` object. Can we refactor all the code in this file to just use that object to serialize/deserialize the query params?
___
*[pkg/ui/src/views/cluster/containers/timescale/index.tsx, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43913#-LyaZnd57xpXLyXuJOrT:-LyaZnd57xpXLyXuJOrU:b-iriuqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/568aa83e9bfbd9bbb13006face814d1caa3ff5b0/pkg/ui/src/views/cluster/containers/timescale/index.tsx#L206)):*
> ```XML
>     router.push({
>       pathname: location.pathname,
>       search: `${removeURLParameters(location, ["start", "end"])}&start=${start}&end=${moment.utc(dateEnd).format("X")}`,
> ```

@elkmaster can you add some tests to reproduce the bug and verify the fix?


<!-- Sent from Reviewable.io -->

vladlos committed:
ui: refactor query

refactored removeURLParameters function

Release note (ui): none

vladlos committed:
ui: query test

Added tests for queries

Release note (ui): none

dhartunian reviewed (changes requested):
> Reviewed 2 of 2 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43913)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @elkmaster)

___
*[pkg/ui/src/util/query.spec.ts, line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43913#-LyoxMLK6IgvpKGBpExR:-LyoxMLK6IgvpKGBpExS:b-b39hie) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a55ca75867452c5a3c003e32194b1f1ecdbc854/pkg/ui/src/util/query.spec.ts#L11)):*
> ```XML
> // licenses/APL.txt.
> 
> import { assert } from "chai";
> ```

Thanks for the tests, these are great!
___
*[pkg/ui/src/views/cluster/containers/timescale/index.tsx, line 202 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43913#-Lyowc_pDf8ObA-gedtb:-Lyowc_pDf8ObA-gedtc:bd0kx2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc58bacccb02209d583c619efb565e99879f2387/pkg/ui/src/views/cluster/containers/timescale/index.tsx#L202)):*
> ```XML
>     const { router, location } = this.props;
>     const seconds = duration.clone().asSeconds();
>     const end = dateEnd.clone();
> ```

Why do we call `clone()` on the durations?
___
*[pkg/ui/src/views/cluster/containers/timescale/index.tsx, line 206 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43913#-LyoQl5BB2a-HGcqzA2C:-LyoQl5C4TzZQs3zgFHj:blax72y) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc58bacccb02209d583c619efb565e99879f2387/pkg/ui/src/views/cluster/containers/timescale/index.tsx#L206)):*
> ```XML
>     router.push({
>       pathname: location.pathname,
>       search: `${removeURLParameters(location, ["start", "end"])}&start=${start}&end=${moment.utc(dateEnd).format("X")}`,
> ```

After looking at this again, I don't think the introduction of `removeURLParameters` is necessary. All direct string manipulation of the query string should be avoided.

Can we accomplish all this like this?

```
    const params = new URLSearchParams(location.search);
    params.set("start", start);
    params.set("end", moment.utc(dateEnd).format("X"));
    router.push({
      pathname: location.pathname,
      search: `?${params.toString()}`,
    });
```

If you agree, I'd prefer this approach as it leans on existing DOM APIs without introducing any new functions and clarifies the intent which is simply to set new start and end dates in the query string.


<!-- Sent from Reviewable.io -->

vladlos committed:
ui: chart selection query

set start and end params to query on existing DOM APIs without introducing removeURLParameters function

Release note (ui): none

dhartunian commented on pkg/ui/src/util/query.ts:
```diff
@@ -45,3 +46,9 @@ export function queryByName(location: Location, key: string) {
   return urlParams.get(key);
 }
+
+export function removeURLParameters(location: Location, removeParams: string[]) {
```
Remove this function, it is now unused.

dhartunian commented on pkg/ui/src/util/query.ts:
```diff
@@ -9,6 +9,7 @@
 // licenses/APL.txt.
 
 import { Location } from "history";
+import _ from "lodash";
```
Remove unused import please

dhartunian reviewed (approved):
Thanks! LGTM.

dhartunian commented:
bors r+

dhartunian commented:
@elkmaster can you rebase on master? I think that will help tests pass.

dhartunian commented:
@elkmaster can you take a look at `make ui-lint` output? I think that's what's preventing your PR from passing TeamCity.

```
src/util/query.spec.ts(22,3): error TS2322: Type '{ pathname: string; search: string; action: string; key: string; state: string; query: { charts: string; end: string; start: string; }; }' is not assignable to type 'Location<{}>'.
  Object literal may only specify known properties, and 'action' does not exist in type 'Location<{}>'.
src/util/query.spec.ts(49,77): error TS2339: Property 'query' does not exist on type 'Location<{}>'.
src/util/query.spec.ts(50,74): error TS2339: Property 'query' does not exist on type 'Location<{}>'.
src/util/query.spec.ts(51,90): error TS2339: Property 'query' does not exist on type 'Location<{}>'.
src/util/query.spec.ts(52,74): error TS2339: Property 'query' does not exist on type 'Location<{}>'.
src/util/query.spec.ts(53,77): error TS2339: Property 'query' does not exist on type 'Location<{}>'.
src/util/query.ts(32,29): error TS2339: Property 'keys' does not exist on type 'URLSearchParams'.
make: *** [ui-lint] Error 2
```

dhartunian commented on pkg/ui/src/views/cluster/containers/timescale/index.tsx:
```diff
@@ -182,6 +182,9 @@ class TimeScaleDropdown extends React.Component<TimeScaleDropdownProps, {}> {
     const query = queryToObj(location, dataType, timestamp);
+    queryToObj(location, "test", "test");
+    queryToObj(location, "test", undefined);
+    queryToObj(location, undefined, undefined);
```
@elkmaster testing something? I think this should be removed.

vladlos committed:
ui: custom chart selection

Removed unused removeUrlParameters function and import

Release note(ui): none

dhartunian commented:
bors r+

dhartunian commented:
bors r+

dhartunian committed:
ui: fix failing queryByName test and delete failing test

The `queryByName` test was comparing an incorrect value
and the `queryToObj` test could not pass as written.

I have deleted the test for now since I've also opened the issue

Release note: none

dhartunian commented:
@elkmaster I've enabled automatic `ui-test` runs in TeamCity so the merge failed. I've pushed some changes to your tests to merge this change.

dhartunian commented:
bors r+

dhartunian commented:
bors r+

dhartunian commented:
bors r+

dhartunian commented:
bors r+

dhartunian commented:
bors r+
